dec2flt: Clean up float parsing modules

tgross35 opened the pull request:
This is the first portion of my work adding support for parsing and printing `f16`. Changes in `float.rs` replace the magic constants with expressions and add some use of generics to better support the new float types. Everything else is related to documentation or naming; there are no functional changes in this PR.

This can be reviewed by commit.

tgross35 committed:
float: Update some constants to `pub(crate)`

These constants can be useful outside of their current module. Make them
`pub(crate)` to allow for this.

tgross35 committed:
dec2flt: Refactor float traits

A lot of the magic constants can be turned into expressions. This
reduces some code duplication.

Additionally, add traits to make these operations fully generic. This
will make it easier to support `f16` and `f128`.

tgross35 committed:
dec2flt: Refactor the fast path

This is just a bit of code cleanup to make use of returning early.

tgross35 committed:
dec2flt: Rename `Decimal` to `DecimalSeq`

This module currently contains two decimal types, `Decimal` and
`Number`. These names don't provide a whole lot of insight into what
exactly they are, and `Number` is actually the one that is more like an
expected `Decimal` type.

In accordance with this, rename the existing `Decimal` to `DecimalSeq`.
This highlights that it contains a sequence of decimal digits, rather
than representing a base-10 floating point (decimal) number.

Additionally, add some tests to validate internal behavior.

tgross35 committed:
dec2flt: Update documentation of existing methods

Fix or elaborate existing float parsing documentation. This includes
introducing a convention that should make naming more consistent.

tgross35 committed:
dec2flt: Rename `Number` to `Decimal`

The previous commit renamed `Decimal` to `DecimalSeq`. Now, rename the
type that represents a decimal floating point number to be `Decimal`.

Additionally, add some tests for internal behavior.

tgross35 committed:
dec2flt: Rename fields to be consistent with documented notation

Noratrieb commented:
I'm by no means an expert in float parsing, but this change is mostly mechanical, so I do feel comfortable approving it. I looked through each commit and the changes seemed reasonable. For the float trait refactor I did take a closer look but a mistake there could have definitely slipped by (though I guess that's always true), but I assume you also ran the detailed float parsing tests (I think they don't run in CI) to ensure it still works as intended.

@bors r+

bors commented:
:pushpin: Commit 89f2066203acdcabf690c9d2397c94c9feed3cb9 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 89f2066203acdcabf690c9d2397c94c9feed3cb9 -->
<!-- homu: {"type":"Approved","sha":"89f2066203acdcabf690c9d2397c94c9feed3cb9","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
> but I assume you also ran the detailed float parsing tests (I think they don't run in CI) to ensure it still works as intended.

Just to confirm, these do still pass. We run a few minutes of edge cases in CI but the `test-float-parse` crate needs to be run locally to get the full few hours of exhaustive (`f32`) / fuzz (`f64`) tests.

> I'm by no means an expert in float parsing

I don't think anyone here is, I'm doing a lot of reverse engineering and the people who authored this implementation don't seem to be active anymore either 😞 but thanks for the review!

tgross35 commented:
_Agh_ the exhaustive tests pass but one of the newer edge case tests failed when I tested with a rebase. Looks like I need to adjust something here.

@bors r-

bors commented:
:umbrella: The latest upstream changes (presumably #134286) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
Finally circling back to this; I had a mistake in the formula for `SMALLEST_POWER_OF_TEN`, that value can't easily be computed with any of our primitive types. So I:

- Switched back to the magic numbers for that constant
- Added a test for the computed numbers against the original magic numbers (also a couple of other tests from the failure)
- Renamed some of the constants for consistency with our many other rust-lang `Float` traits

And test-float-parse completes successfully again 🎉. Since nothing functional changed outside of the fixed constant,

@bors r=Noratrieb

bors commented:
:pushpin: Commit 37e223ccaa1d0036540e6e3de5d4ed4b721afd60 has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 37e223ccaa1d0036540e6e3de5d4ed4b721afd60 -->
<!-- homu: {"type":"Approved","sha":"37e223ccaa1d0036540e6e3de5d4ed4b721afd60","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
